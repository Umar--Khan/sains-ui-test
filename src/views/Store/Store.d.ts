export interface Product {
  productId: string;
  sku: string;
  title: string;
  price: number;
  image: string;
}

export interface BasketProduct {
  quantity: number;
  productDetail: Product;
}
