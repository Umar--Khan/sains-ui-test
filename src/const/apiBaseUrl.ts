export const apiBaseUrl = 'https://jsainsburyplc.github.io/front-end-test/';
