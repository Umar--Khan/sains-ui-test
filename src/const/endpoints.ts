const endpoints = {
  getProducts: 'products.json',
};

export default endpoints;
