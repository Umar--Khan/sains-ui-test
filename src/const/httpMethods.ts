export enum HTTP_METHODS {
  GET = 'GET',
}
