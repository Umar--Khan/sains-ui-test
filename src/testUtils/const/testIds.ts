export const headerTestIds = {
  numberInBasket: 'number-in-basket',
};

export const storeTestIds = {
  productCard: 'store-product-card',
};

export const basketTestIds = {
  reviewCard: 'basket-review-card',
  emptyBasketMessage: 'empty-basket-message',
};
