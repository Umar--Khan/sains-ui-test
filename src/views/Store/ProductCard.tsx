import Card from 'src/shared/Card/Card';
import { Button } from 'src/shared/Buttons';
import { Product } from './Store.d';
import styled from 'styled-components';

export interface ProductCardProps {
  product: Product;
  handleOnClick: (arg: Product) => void;
}

const Sku = styled.span`
  font-size: 0.8rem;
  font-weight: 400;
`;

const Img = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const ProductCard = ({ product, handleOnClick, ...rest }: ProductCardProps) => (
  <Card key={product.sku} {...rest}>
    <Img src={product.image} alt={product.title} />
    <p>
      {product.title}, <Sku>{product.sku}</Sku>
    </p>
    <p>&#163;{product.price}</p>
    <Button onClick={() => handleOnClick(product)}>Add to basket</Button>
  </Card>
);

export default ProductCard;
