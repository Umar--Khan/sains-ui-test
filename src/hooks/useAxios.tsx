import { useCallback, useEffect, useState } from 'react';

import axiosConfig from 'src/axios/axiosConfig';
import { HTTP_METHODS } from 'src/const/httpMethods';

export interface UseResourceProps {
  url: string;
  method?: HTTP_METHODS;
  body?: Record<string, any>;
  headers?: Record<string, string>;
}

const useAxios = ({
  url,
  method = HTTP_METHODS.GET,
  body,
  headers,
  ...rest
}: UseResourceProps) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState(null);

  const fetchData = useCallback(async (): Promise<void> => {
    setIsLoading(true);
    try {
      const res = await axiosConfig({
        url,
        method,
        data: body,
        headers,
        ...rest,
      });
      if (res.data) {
        setData(res.data);
      }
    } catch (error) {
      setError(true);
      setErrorMessage(error?.response?.data);
    }
    setIsLoading(false);
    // eslint-disable-next-line
  }, [url]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return { data, error, isLoading, errorMessage };
};

export default useAxios;
