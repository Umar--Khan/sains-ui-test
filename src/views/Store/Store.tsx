import { useEffect, useState, useContext } from 'react';

import useAxios from 'src/hooks/useAxios';
import { BasketContext } from 'src/context/BasketContext';
import endpoints from 'src/const/endpoints';
import { storeTestIds } from 'src/testUtils/const/testIds';
import { Product, BasketProduct } from './Store.d';
import ProductCard from './ProductCard';

const STORE_ERROR_MESSAGE =
  'We are having problems loading the store. Please check back later.';

const Store = () => {
  const [products, setProducts] = useState<Product[] | null>(null);
  const { data, error, isLoading } = useAxios({
    url: endpoints.getProducts,
  });

  const { basketLocal, setBasketLocal } = useContext(BasketContext);

  useEffect(() => {
    setProducts(data);
  }, [data]);

  const handleAddToBasket = (newProduct: Product) => {
    const isProductInBasket = basketLocal.some(
      ({ productDetail: { productId } }: BasketProduct) =>
        productId === newProduct.productId,
    );

    if (!isProductInBasket) {
      return setBasketLocal([
        ...basketLocal,
        { productDetail: newProduct, quantity: 1 },
      ]);
    }

    const updatedBasket = basketLocal.map(
      ({ productDetail, quantity }: BasketProduct) => {
        if (productDetail.productId === newProduct.productId) {
          return { productDetail, quantity: quantity + 1 };
        }
        return { productDetail, quantity };
      },
    );

    setBasketLocal(updatedBasket);
  };

  if (isLoading) return <div>Loading...</div>;
  if (error) return <div>{STORE_ERROR_MESSAGE}</div>;

  return (
    <>
      {products ? (
        products.map(product => (
          <ProductCard
            product={product}
            handleOnClick={handleAddToBasket}
            key={product.productId}
            data-testid={storeTestIds.productCard}
          />
        ))
      ) : (
        <h1>{STORE_ERROR_MESSAGE}</h1>
      )}
    </>
  );
};
export default Store;
