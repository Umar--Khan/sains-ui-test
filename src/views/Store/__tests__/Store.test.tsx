import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import useAxios from 'src/hooks/useAxios';
import { mockedGetProducts } from 'src/testUtils/mocks/getProducts';
import { storeTestIds } from 'src/testUtils/const/testIds';
import Store from '../Store';

jest.mock('src/hooks/useAxios');

describe('<Store />', () => {
  describe('<Store /> Snapshots', () => {
    it('Setup', () => {
      useAxios.mockReturnValue({
        data: mockedGetProducts,
        isLoading: false,
        error: false,
      });

      renderer.create(<Store />);
    });

    it('Should render snapshot correctly', () => {
      useAxios.mockReturnValue({
        data: mockedGetProducts,
        isLoading: false,
        error: false,
      });

      const component = renderer.create(<Store />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('Should render error snapshot correctly', () => {
      useAxios.mockReturnValue({
        data: [],
        isLoading: false,
        error: true,
      });

      const component = renderer.create(<Store />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('Should render loading snapshot correctly', () => {
      useAxios.mockReturnValue({
        data: [],
        isLoading: true,
        error: false,
      });

      const component = renderer.create(<Store />);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it('Should render correct amount of products', () => {
    useAxios.mockReturnValue({
      data: mockedGetProducts,
    });
    const { getAllByTestId } = render(<Store />);
    const productCards = getAllByTestId(storeTestIds.productCard);

    expect(productCards).toHaveLength(mockedGetProducts.length);
  });
});
