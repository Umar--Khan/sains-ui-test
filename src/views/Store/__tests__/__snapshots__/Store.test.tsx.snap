// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Store /> <Store /> Snapshots Should render error snapshot correctly 1`] = `
<div>
  We are having problems loading the store. Please check back later.
</div>
`;

exports[`<Store /> <Store /> Snapshots Should render loading snapshot correctly 1`] = `
<div>
  Loading...
</div>
`;

exports[`<Store /> <Store /> Snapshots Should render snapshot correctly 1`] = `
<div
  className="sc-eCssSg jBlPHF"
>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Russell Hobbs Pennine Illuminating S Steel Kettle 20444"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/9805244_R_SET?w=110&h=130"
    />
    <p>
      Russell Hobbs Pennine Illuminating S Steel Kettle 20444
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        549/3179
      </span>
    </p>
    <p>
      £
      24.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Challenge White Desk Fan - 12 Inch"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/4251824_R_SET?w=110&h=130"
    />
    <p>
      Challenge White Desk Fan - 12 Inch
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        425/1824
      </span>
    </p>
    <p>
      £
      14.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Plain Vest Top 12 BLACK"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/tuc132404445_R_SET?w=110&h=130"
    />
    <p>
      Plain Vest Top 12 BLACK
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        tuc132404445
      </span>
    </p>
    <p>
      £
      16
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Wilson Roger Federer 27 Inch Adult Tennis Racket"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/4049641_R_SET?w=110&h=130"
    />
    <p>
      Wilson Roger Federer 27 Inch Adult Tennis Racket
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        404/9641
      </span>
    </p>
    <p>
      £
      24.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="My Little Pony Cutie Mark Rainbow Dash"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/5399235_R_SET?w=110&h=130"
    />
    <p>
      My Little Pony Cutie Mark Rainbow Dash
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        539/9235
      </span>
    </p>
    <p>
      £
      24.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Sim Free iPhone X 256GB Mobile Phone- Space Grey"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/7543412_R_SET?w=110&h=130"
    />
    <p>
      Sim Free iPhone X 256GB Mobile Phone- Space Grey
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        754/3412
      </span>
    </p>
    <p>
      £
      1109
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Assassin's Creed Origins PS4 Game"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/7067462_R_SET?w=110&h=130"
    />
    <p>
      Assassin's Creed Origins PS4 Game
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        706/7462
      </span>
    </p>
    <p>
      £
      30.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
  <div
    className="sc-bdfBwQ fvoACg"
    data-testid="store-product-card"
  >
    <img
      alt="Duracell Supreme 750 mAh Rechargeable AAA Batteries - 4 Pack"
      className="sc-hKgILt eDdGvz"
      src="https://media.4rgos.it/s/Argos/7426735_R_SET?w=110&h=130"
    />
    <p>
      Duracell Supreme 750 mAh Rechargeable AAA Batteries - 4 Pack
      , 
      <span
        className="sc-dlfnbm bQGhrs"
      >
        742/6735
      </span>
    </p>
    <p>
      £
      8.99
    </p>
    <button
      className="sc-gsTCUz kiZIlZ"
      onClick={[Function]}
    >
      Add to basket
    </button>
  </div>
</div>
`;
